Refactor and clean up Clipper admin python package

dcrankshaw opened the pull request:
This pull request proposes a new implementation of the Clipper admin Python package for managing and deploying a Clipper cluster. The new implementation provides a new API and an internal re-organization of the code. Aside from general cleanup, this implementation makes several improvements. Details and requests for feedback will be coming soon.

feynmanliang committed:
Initial refactor of Clipper admin package.

This commit factors out Docker specific functionality behind a container
manager interface. It provides a working implementation of the container
manager interface with Docker, and it provides a mostly complete
implementation using k8s. It also factors out framework-specific
deployment methods into their own modules and removes the model
container need to mount local filesystem paths into the Docker
container by packaging model data into the Docker images themselves.

The integration tests currently pass with the Docker container manager
implementation but not the k8s implementation.

dcrankshaw committed:
ported link model functionality

dcrankshaw committed:
integration tests actually passing

dcrankshaw committed:
changed integration tests to use k8s_container_manager

dcrankshaw commented:
@feynmanliang This is finally ready for another round of work on the k8s support. I'd like to treat this branch as a "feature branch" and submit PRs against it until the package is fully ready to be merged.

### Some context
The package has been re-organized pretty substantially and the new package is in `clipper_admin_v2/`. The relevant piece for you is that the container management functionality has been factored out into a [`ContainerManager`](https://github.com/dcrankshaw/clipper/blob/clipper_admin_v2/clipper_admin_v2/clipper_admin/container_manager.py) interface and there are currently two implementations: a pure Docker implementation in `clipper_admin_v2/clipper_admin/docker/docker_container_manager.py` and the K8s implementation you wrote in `clipper_admin_v2/clipper_admin/k8s/k8s_container_manager.py`.

### TODOs for you
[k8s_container_manager.py](https://github.com/dcrankshaw/clipper/blob/clipper_admin_v2/clipper_admin_v2/clipper_admin/k8s/k8s_container_manager.py) should be the only file you have to modify, I stuck some TODOs in there for you. I tried to document what the semantics of each function should be, but you can also look at how [`clipper_admin.py`](https://github.com/dcrankshaw/clipper/blob/clipper_admin_v2/clipper_admin_v2/clipper_admin/clipper_admin.py) uses the ContainerManager interface to see what the functions should do.

If you have any feedback or ideas on how to improve the rest of the package, feel free to implement them or make suggestions, but don't feel obligated to.

### Testing
You'll know that your implementation is correct and complete when the integration tests pass with the Kubernetes container manager. Note that you'll need to rebuild the Clipper docker containers locally before running the tests.

You can run them with

```
./bin/build_docker_images.sh
./bin/run_unittests -i
```

I've modified the integration tests to use the k8s container manager, but if you'd like to see how they run with just Docker, you can change [line 22](https://github.com/dcrankshaw/clipper/blob/clipper_admin_v2/integration-tests/test_utils.py#L22)  in `test_utils.py` to the following to run them with the Docker container manager.

```py
SERVICE = "docker"
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/567/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/568/
Test FAILed.

feynmanliang commented:
I'll look at this today./

feynmanliang commented:
The commit history on the k8s work is lost (presumably because you did a merge against `ucbrise:develop` for 9416786). We could have kept that if this PR used `feynmanliang:k8s` as the base, but since some merge conflicts were resolved the additional commits don't cleanly cherrypick. I don't think it's worth the effort trying to recover them, so lets leave this as is.

feynmanliang commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,87 @@
+        return
+
+    @abc.abstractmethod
+    def add_replica(self, name, version, input_type, repo):
```
Why not `set_number_replicas` to improve referential transparency and also support decrementing?

feynmanliang replied:
Also, why do we need anything more than just the `name` if we're operating on an already deployed model?

dcrankshaw replied:
Those are both good points. The reasons for both choices are historical.

feynmanliang commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,87 @@
+    # when  calling `add_replica()`. In this case, the two methods
+    # would require different behavior.
+    @abc.abstractmethod
+    def deploy_model(self, name, version, input_type, repo):
```
Not sure I understand the comment, but yes you can start a Deployment with 0 replicas and modify that later. However, you'll need this in the superclass because running on K8s requires first creating the Deployment to manage rolling updates/restarting pods/etc whereas when running locally there is no Deployment to create.

Also, in the old CM, this method registers the model with clipper (`self._publish_new_model`). I suspect you'll want to do that here as well (it's currently missing in the Docker CM).

feynmanliang commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,488 @@
+    pass
+
+
+def start_clipper(cm):
```
Any reason why this isn't encapsulated in a class? Seems weird to have the end user maintain the reference to the `ContainerManager`, manage its lifecycle, and pass it into all these functions.

dcrankshaw replied:
You're right, that does seem weird. At first I was thinking that plain functions would be cleaner and you could potentially wrap this in a CLI, but given that there's still this container management state that needs to be tracked, we should create a class and track it for the user.

feynmanliang commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,87 @@
+        self.public_hostname = clipper_public_hostname
+
+    @abc.abstractmethod
+    def start_clipper(self):
```
Seems strange that the `cm` is responsible for starting clipper, but `cm`'s constructor takes Clipper hostname as an argument (which isn't known until after Clipper is started).

I would've made`start_clipper` concrete and have it delegate to an abstract `_start_clipper` that is overridden by implementations of the abc which returns the hostname, then have `start_clipper` (this method) set `self.public_hostname` with the return.

feynmanliang commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,488 @@
+    req_json = json.dumps({
+        "name": name,
+        "input_type": input_type,
+        "default_output": default_output,
```
This missing a field

```
ClipperException: Received error status code: 400 and message: Error parsing JSON: JSON object does not have required key: candidate_model_names. Expected JSON schema:
  {
   "name" := string,
   "candidate_model_names" := [string],
   "input_type" := "integers" | "bytes" | "floats" | "doubles" | "strings",
   "default_output" := string,
   "latency_slo_micros" := int
  }
```

feynmanliang commented on clipper_admin_v2/README.rst:
```diff
@@ -0,0 +1,132 @@
++ Joey Gonzalez (@jegonzal)
++ Corey Zumar (@Corey-Zumar)
++ Nishad Singh (@nishadsingh1)
++ Alexey Tumanov (@atumanov)
```
Add me?

feynmanliang commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,87 @@
+
+# Constants
+
+CLIPPER_QUERY_PORT = 1337
```
These must not be hardcoded for k8s compatibility, where there is no guarantee of the actual port bound by a service. Since we don't ever use these without having a reference to a CM, consider making this a property of a running CM instead

dcrankshaw replied:
Got it. That makes sense.

feynmanliang commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,488 @@
+
+    logger.info("Pushing model Docker image to {}".format(repo))
+    docker_client.images.push(repository=repo)
+    cm.deploy_model(name, version, input_type, repo, num_replicas=num_replicas)
```
This fails because `num_replicas` is not part of CM's public interface (it is provided in the Docker CM). If we want to expose it, we should add it to the CM's public API.

feynmanliang commented on clipper_admin_v2/clipper_admin/k8s/k8s_container_manager.py:
```diff
@@ -0,0 +1,240 @@
+        config.load_kube_config()
+        self._k8s_v1 = client.CoreV1Api()
+        self._k8s_beta = client.ExtensionsV1beta1Api()
+        self.redis_port = redis_port
```
These are duplicated across both CM implementations. Since redis and a registry are both mandatory for a CM to do things (e.g. deploy docker images), this could be initialized in the superclass (via the `super` constructor and `**kwargs`).

feynmanliang committed:
Implements add_replica

feynmanliang committed:
Implements get_logs

feynmanliang committed:
Takes model_name argument in register_application

feynmanliang committed:
Removes print

feynmanliang commented:
I sent you a PR with the TODOs (and some of the comments ^^) at https://github.com/dcrankshaw/clipper/pull/3

dcrankshaw commented:
Awesome. Thanks for the comments. I know that the commit history was lost. I wanted to rebase this PR on the latest version of `develop` and there were a ton of merge conflicts that I had to keep addressing over and over. I made the choice to blow away commit history to make the rebase a lot cleaner. Sorry about any inconvenience.

dcrankshaw committed:
minor fixups

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/591/
Test FAILed.

feynmanliang commented:
Oops I think my PR had a problem: https://github.com/dcrankshaw/clipper/pull/3/files#diff-6272c59650820d22a2a09dfd5eba47f3L123 shouldn't have been removed

dcrankshaw commented:
Got it. I'll add it back in. I fixed a couple other problems before merging as well, so no worries.

dcrankshaw committed:
got clipper admin short integration tests passing again

dcrankshaw committed:
clipper admin tests passing on docker

dcrankshaw committed:
integration tests passing with docker

dcrankshaw committed:
fixed a bunch of k8s issues

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/592/
Test FAILed.

dcrankshaw committed:
more k8s work

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/595/
Test FAILed.

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+    pass
+
+
+class ClipperConnection(object):
```
Can we come up with a better name than ClipperConnection?

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+                except RequestException as e:
+                    logger.info("Clipper still initializing.")
+                    time.sleep(1)
+            logger.info("Clipper is running")
```
Does `start_clipper` need to call `self.connect()`?

dcrankshaw replied:
Addressed

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+            logger.info(e.msg)
+            return False
+
+    def connect(self):
```
Let's set a boolean here that prevents users from calling other methods if they haven't connected yet. This bool can be set in `start_clipper` and in `connect`, and then we can provide nice error messages.

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+            logger.error(msg)
+            raise ClipperException(msg)
+
+    def _resolve_docker_repo(self, image, registry):
```
this is out of date, I will be changing this.

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+            repo = "{registry}/{image}".format(registry=registry, image=image)
+        return repo
+
+    def build_and_deploy_model(self,
```
Extract the `build` part from this function and make a public API method that let's users build a model container and push to a registry without deploying to Clipper.

dcrankshaw replied:
Rename `build_and_deploy_model` to a shorter and cleaner name

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+            docker_client = docker.from_env()
+        logger.info(
+            "Building model Docker image at {}".format(model_data_path))
+        docker_client.images.build(path=model_data_path, tag=repo)
```
Delete the dockerfile we just wrote to keep filesystem clean

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+                     num_replicas=1):
+        version = str(version)
+
+        with open(model_data_path + '/Dockerfile', 'w') as f:
```
rename the Dockerfile to a random string to avoid overwriting existing dockerfiles and set the `dockerfile` argument in `docker_client.images.build`

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+                     name,
+                     version,
+                     input_type,
+                     repo,
```
rename `repo` to `image`

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+        version = str(version)
+        return self.cm.get_num_replicas(name, version)
+
+    def set_num_replicas(self, name, version=None, num_replicas=1):
```
Does this method name adequately convey that replicas will be added or removed (which is a fairly heavyweight operation)?

dcrankshaw replied:
I think it's fine

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+        version = str(version)
+        return self.cm.get_num_replicas(name, version)
+
+    def set_num_replicas(self, name, version=None, num_replicas=1):
```
Remove the default value from `num_replicas` and force the user to be explicit

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+        return self.cm.get_query_addr()
+
+    def stop_deployed_models(self):
+        self.cm.stop_models()
```
add a `def stop_models(self, names)` method that takes a list of model names and stops all versions of each of those models

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+            if m["is_current_version"]:
+                current_version = m["model_version"]
+        assert current_version is not None
+        self.cm.stop_models(
```
Change the container manager `stop_models` method to take a dict of <model_name: list(model_version)> pairs. For each model. this will stop each version in that model's list.

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,86 @@
+        return
+
+    @abc.abstractmethod
+    def deploy_model(self, name, version, input_type, repo):
```
rename repo to image

dcrankshaw commented on clipper_admin_v2/clipper_admin/container_manager.py:
```diff
@@ -0,0 +1,86 @@
+    def stop_clipper(self):
+        pass
+
+    def get_registry(self):
```
delete this method

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+
+class DockerContainerManager(ContainerManager):
+    def __init__(self,
+                 clipper_public_hostname,
```
This argument doesn't make sense, it should be set in `start_clipper` or provided in `connect()`

dcrankshaw replied:
In retrospect I don't think this comment makes sense. Any docker specific arguments should be provided to init, so that `start_clipper` and `connect` can cleanly inherit from the generic `ContainerManager` interface. I'm leaving this as is.

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+                 redis_port=6379,
+                 registry=None,
+                 registry_username=None,
+                 registry_password=None,
```
Delete registry related arguments

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+                 clipper_management_port=1338,
+                 clipper_rpc_port=7000,
+                 redis_ip=None,
+                 redis_port=6379,
```
Move connection-related arguments to be set in `start_clipper` and `connect`

dcrankshaw replied:
See above comment

dcrankshaw commented on clipper_admin_v2/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,539 @@
+                     version,
+                     input_type,
+                     model_data_path,
+                     base_image,
```
Add an additional metadata field that will be provided as a string environment variable to each container when run. This can be used to pass additional model metadata (such as object store addresses, etc) that need to be resolved at runtime and can't be put into the image.

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+            },
+            "detach": True,
+        }
+        self.extra_container_kwargs.update(container_args)
```
This overwrites any labels the user provided. We should extend the labels list and overwrite the network.

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+            self.redis_ip = "redis"
+            self.docker_client.containers.run(
+                'redis:alpine',
+                "redis-server --port %d" % self.redis_port,
```
This should always start redis on the default port, and just remap 6379 to the provided redis port in the `ports` argument.

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+        cmd = "--redis_ip={redis_ip} --redis_port={redis_port}".format(
+            redis_ip=self.redis_ip, redis_port=self.redis_port)
+        self.docker_client.containers.run(
+            'clipper/management_frontend:latest',
```
These should not pull the `latest` tagged versions from docker hub.

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+            },
+            **self.extra_container_kwargs)
+        self.docker_client.containers.run(
+            'clipper/query_frontend:latest',
```
change version tag

dcrankshaw commented on clipper_admin_v2/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,242 @@
+                ":")
+            if model_name is not None and model_name == c_name:
+                if keep_version is None or keep_version != c_version:
+                    c.stop()
```
Well I'm going to rewrite this function anyway, but this misses the case where model_name is None and we want to remove all model containers

dcrankshaw commented on clipper_admin_v2/clipper_admin/deployers/python.py:
```diff
@@ -0,0 +1,131 @@
+                    func,
+                    default_output="None",
+                    version=1,
+                    slo_micros=100000,
```
Bump default slo_micros to 3 seconds

dcrankshaw replied:
Done

dcrankshaw commented on clipper_admin_v2/clipper_admin/deployers/python.py:
```diff
@@ -0,0 +1,131 @@
+                          version,
+                          input_type,
+                          func,
+                          base_image="clipper/python-closure-container",
```
This should have a version tag associated with it

dcrankshaw commented on clipper_admin_v2/clipper_admin/deployers/python.py:
```diff
@@ -0,0 +1,131 @@
+    serialization_dir = save_python_function(name, func)
+    logger.info("Python closure saved")
+    # Deploy function
+    deploy_result = clipper_conn.deploy_model(name, version, input_type,
```
I don't think `deploy_model` returns anything anymore?

dcrankshaw commented on clipper_admin_v2/clipper_admin/deployers/pyspark.py:
```diff
@@ -0,0 +1,151 @@
+                    sc,
+                    default_output="None",
+                    version=1,
+                    slo_micros=100000,
```
bump default slo

dcrankshaw replied:
Done

dcrankshaw reviewed (commented):
Comments from joint review with @Corey-Zumar

dcrankshaw committed:
k8s stop and delete seems to be working

dcrankshaw committed:
k8s starting and stopping clipper works. Problem with model container naming conventions

dcrankshaw committed:
got model and app names fixed

dcrankshaw committed:
Successfully ran clipper kubernetes end to end test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/602/
Test FAILed.

dcrankshaw committed:
started fixing rest of unit tests

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/603/
Test FAILed.

dcrankshaw committed:
All integration tests pass with docker

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/619/
Test FAILed.

dcrankshaw committed:
update versions

dcrankshaw committed:
updated tf cifar dockerfile

dcrankshaw committed:
ported module dep saver to clipper_admin_v2

dcrankshaw committed:
integration tests passing

dcrankshaw committed:
added k8s test and partway through addressing code review comments

dcrankshaw committed:
addressed rest of code review comments

dcrankshaw committed:
all integration tests passing

dcrankshaw committed:
worked on doc skeleton and port to rst

Corey-Zumar commented:
@dcrankshaw Is this ready for review again?

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/637/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/638/
Test FAILed.

dcrankshaw committed:
Added lots of documentation

dcrankshaw committed:
more docs

dcrankshaw committed:
renamed k8s to kubernetes and deleted clipper_admin

dcrankshaw committed:
renamed clipper_admin_v2 to clipper_admin

dcrankshaw committed:
updated readme

dcrankshaw committed:
update docs and format code

dcrankshaw committed:
added manifest and fixed version

dcrankshaw committed:
added cache size arg

dcrankshaw commented:
@Corey-Zumar This is ready for a review. There are a couple documentation loose ends I need to wrap up (the quickstart in the README and the tutorial and examples), but other than that this should be ready.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/649/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/651/
Test FAILed.

dcrankshaw committed:
format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/653/
Test FAILed.

shaneknapp commented:
test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/654/
Test FAILed.

shaneknapp commented:
test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/655/
Test FAILed.

dcrankshaw committed:
removed static names for clipper docker containers

Corey-Zumar commented on PySparkContainerDockerfile:
```diff
@@ -1,9 +1,7 @@
 ARG CODE_VERSION=0.2-rc1
```
This should be `develop`, right?

dcrankshaw replied:
Fixed

Corey-Zumar commented on PyClosureContainerDockerfile:
```diff
@@ -0,0 +1,14 @@
+ARG CODE_VERSION=0.2-rc1
```
`develop`

dcrankshaw replied:
Fixed

Corey-Zumar commented on bin/run_unittests.sh:
```diff
@@ -138,10 +138,13 @@ function run_integration_tests {
-  python ../integration-tests/many_apps_many_models.py 2 3
+  python ../integration-tests/kubernetes_integration_test.py
+  # TODO: uncomment
+  # python ../integration-tests/deploy_R_models.py
```
Assuming this was commented out due to impending R support?

dcrankshaw replied:
Decided to delete the weird R/Python container. I deleted the commented out line.

Corey-Zumar commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,256 @@
+            The public hostname or IP address at which the Clipper Docker
+            containers can be accessed via their exposed ports. If running the containers
+            on a network other than "host", you may need to set this explicitly for ``ClipperConnection``
+            access the Clipper containers.
```
Nit: Typo, should be `...you may need to set this explicitly for ``ClipperConnection`` **to** access the Clipper containers`

Corey-Zumar commented on clipper_admin/setup.py:
```diff
@@ -0,0 +1,38 @@
+from setuptools import setup
```
`python setup.py install` fails due to the absence of a `requirements.txt` file (Python 2.7).

dcrankshaw replied:
Dependencies are specified in  `setup.py`. Run `pip install -e /path/to/clipper_admin` instead.

Corey-Zumar commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,256 @@
+        for c in containers:
+            c.stop()
+
+    def stop_all(self):
```
If `redis_ip` was initially `None`, we don't reset it when we stop all containers. As a result, subsequent calls to `start_clipper` fail to start redis.

dcrankshaw replied:
Added an additional field `self.external_redis` that we never mutate outside the constructor:

```
        if redis_ip is None:
            self.external_redis = False
        else:
            self.external_redis = True
```

If `self.external_redis` is False when you call `start_clipper`, we will start a redis instance regardless of whether you've already started and stopped clipper.

Corey-Zumar commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,256 @@
+                ports={'%s/tcp' % self.redis_port: self.redis_port},
+                labels=self.common_labels,
+                **self.extra_container_kwargs)
+            self.redis_ip = redis_container.name
```
In the case where we're using the docker host network, this should be `self.redis_ip = localhost`. Otherwise, the frontend containers fail to connect to redis.

dcrankshaw replied:
See above comment, "host" network is no longer supported.

Corey-Zumar commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,256 @@
+                ports={'%s/tcp' % self.redis_port: self.redis_port},
+                labels=self.common_labels,
+                **self.extra_container_kwargs)
+            self.redis_ip = redis_container.name
```
Also, this seems to be failing in general for other docker networks. The frontend containers cannot find redis using the container name as the ip address.

dcrankshaw replied:
I think this should all work now

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/656/
Test FAILed.

Corey-Zumar commented on clipper_admin/docs/conf.py:
```diff
@@ -42,13 +42,13 @@
-from recommonmark.transform import AutoStructify
-
-source_parsers = {'.md': CommonMarkParser}
+# import recommonmark
```
Remove commented code

dcrankshaw replied:
Given that this is just the config to build the docs, I'd actually rather leave that in if we want to add markdown support back, rather than having to go look it up again in the future.

Corey-Zumar commented on clipper_admin/docs/conf.py:
```diff
@@ -167,15 +168,23 @@
-    }, True)
-    app.add_transform(AutoStructify)
+    app.connect("autodoc-skip-member", skip)
+    # app.add_config_value('recommonmark_config', {
```
Remove commented code

dcrankshaw replied:
See above comment

Corey-Zumar commented on containers/python/python_closure_container.py:
```diff
@@ -3,10 +3,9 @@
 
-sys.path.append(os.path.abspath("/lib/"))
-from clipper_admin import cloudpickle
+# sys.path.append(os.path.abspath("/lib/"))
```
Should this be removed or uncommented?

dcrankshaw replied:
Removed

Corey-Zumar commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -0,0 +1,256 @@
+        else:
+            self.common_labels = {CLIPPER_DOCKER_LABEL: ""}
+
+    def start_clipper(self, query_frontend_image, mgmt_frontend_image,
```
In the case that the docker network is `host`, this fails to forward the proper ports. As a result, requests sent to `localhost:1337/metrics` by `ClipperConnection` fail and `ClipperConnection.start()` never returns.

dcrankshaw replied:
Yeah, turns out setting `docker_network=host` causes some weirdness. I reverted back to creating our own network and we throw an exception if the user tries to set the network to "host". This should work now.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,995 @@
+            logger.error(msg)
+            raise ClipperException(msg)
+
+    def build_and_deploy_model(self,
```
In addition to this method and `deploy_model`, I still think it's a good idea to have a `build_model` method defined within this module that performs only the first operation documented below. This will be useful for data scientists who want to deploy their models without access to (or a need for) a running Clipper instance.

dcrankshaw replied:
Yeah good point. Done.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,995 @@
+
+        _validate_versioned_model_name(name, version)
+
+        with open(model_data_path + '/Dockerfile', 'w') as f:
```
As far as I can tell, couldn't we still potentially be overwriting a dockerfile here? There was a comment indicating that this was fixed, but I'm not sure:

Comment:
"rename the Dockerfile to a random string to avoid overwriting existing dockerfiles and set the `dockerfile` argument in `docker_client.images.build`"

dcrankshaw replied:
Yeah that didn't work with the Python docker SDK. I added a check and raise an exception if a Dockerfile already exists at that location.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,995 @@
+            model_info = self.get_all_models(verbose=True)
+            for m in model_info:
+                if m["is_current_version"]:
+                    version = m["model_version"]
```
Nit: We can insert a break after we've found the most recent version

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,995 @@
+        if version is None:
+            model_info = self.get_all_models(verbose=True)
+            for m in model_info:
+                if m["is_current_version"]:
```
I think this needs to be filtered by the supplied model name. Otherwise, we're setting `version` to be the current version of the last model contained in the dictionary.

dcrankshaw replied:
Oh yeah. Good catch.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,995 @@
+            model_info = self.get_all_models(verbose=True)
+            for m in model_info:
+                if m["is_current_version"]:
+                    version = m["model_version"]
```
This logic needs to be changed in the same way as lines 453-459. We should define a `get_current_model_version(name)` method to minimize duplicate code.

dcrankshaw replied:
Done

Corey-Zumar reviewed (changes requested):
This is awesome - very intuitive API. There are a few functionality issues to address (see comments). 

Notes:

* Issues with DockerContainerManager are preventing full testing of ClipperConnection. I will test the full API once these issues have been addressed. 

* Unless the user looks at the source for logging behavior, they may not have any log output by default. Logging does not work until `logging.basicConfig()` is called. We should call it for the user in some fashion. 

* Resolve all outstanding TODOs or comment indicating that they’ll be addressed in a subsequent PR

dcrankshaw committed:
fixed some bugs

dcrankshaw committed:
format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/657/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/658/
Test FAILed.

dcrankshaw committed:
addressed review comments

dcrankshaw commented:
@Corey-Zumar addressed your review comments. I still have a few TODOs to address and I need to update the tutorial and quickstart, but other than that this is complete and ready for another round of reviews.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/659/
Test FAILed.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        })
+        headers = {'Content-type': 'application/json'}
+        r = requests.post(url, headers=headers, data=req_json)
+        if r.status_code != requests.codes.ok:
```
Else (if the status code equals `requests.codes.ok`), we should log the response text.

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        r = requests.post(url, headers=headers, data=req_json)
+
+        if r.status_code == requests.codes.ok:
+            app_info = r.json()
```
Nit: This should be `model_info`. Also, this method does not log any content in the case that the user attempts to get info corresponding to an unregistered model.

dcrankshaw replied:
Good point. Fixed.

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        if r.status_code == requests.codes.ok:
+            app_info = r.json()
+            if len(app_info) == 0:
+                return None
```
Consider indicating to the user that no such replica exists.

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        r = requests.post(url, headers=headers, data=req_json)
+
+        if r.status_code == requests.codes.ok:
+            app_info = r.json()
```
Nit: `replica_info`

dcrankshaw replied:
Fixed

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+                    model_dict[m["model_name"]].append(m["model_version"])
+                else:
+                    model_dict[m["model_name"]] = [m["model_version"]]
+        self.cm.stop_models(model_dict)
```
It would be nice to log the number and versions of models that were stopped.

dcrankshaw replied:
Logged the model dict (with a pretty printer so it should be readable)

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        Clipper management frontend or Redis has crashed. It can also be called without calling
+        ``connect`` first.
+        """
+        self.cm.stop_all_model_containers()
```
Log info about stopped containers or at least a success message.

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -0,0 +1,1094 @@
+        If you started Redis independently, this will not affect Redis. It can also be called without calling
+        ``connect`` first.
+        """
+        self.cm.stop_all()
```
Log a success message

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/clipper_admin/deployers/python.py:
```diff
@@ -0,0 +1,134 @@
+    clipper_conn.link_model_to_app(name, name)
+
+
+def deploy_python_closure(
```
After invoking `deploy_python_closure(conn, "m7", "1", "ints", func) ` with a connection backed by the docker container manager, I encountered an error with the following back trace:

```
Traceback (most recent call last):
  File "/Users/Corey/Documents/RISE/clipper/clipper_admin/clipper_admin/deployers/check_and_write_deps.py", line 127, in <module>
    conda_dep_fname, pip_dep_fname):
  File "/Users/Corey/Documents/RISE/clipper/clipper_admin/clipper_admin/deployers/check_and_write_deps.py", line 91, in check_solvability_write_deps
    missing_packages = missing_packages_error.pkgs
AttributeError: 'ResolvePackageNotFound' object has no attribute 'pkgs'

17-08-30:02:04:13 INFO     [python.py:127] Python closure saved
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Corey/Documents/RISE/clipper/clipper_admin/clipper_admin/deployers/python.py", line 138, in deploy_python_closure
    force=True)
  File "/Users/Corey/Documents/RISE/clipper/clipper_admin/clipper_admin/clipper_admin.py", line 281, in build_and_deploy_model
    container_registry)
  File "/Users/Corey/Documents/RISE/clipper/clipper_admin/clipper_admin/clipper_admin.py", line 346, in build_model
    docker_file_path = os.path.join(model_data_path, "Dockerfile")
  File "/usr/local/bin/anaconda2/envs/clipper/lib/python2.7/posixpath.py", line 70, in join
    elif path == '' or path.endswith('/'):
AttributeError: 'bool' object has no attribute 'endswith'
```

dcrankshaw replied:
This issue looks like a problem introduced by the latest version of conda. We got a bug report about this in #269.

What version of conda are you on?

dcrankshaw replied:
Can you try rolling back to a 4.3.* version of conda and running again. This is definitely a bug that we should fix, but isn't a new one introduced by this PR and shouldn't block this getting merged.

Corey-Zumar replied:
I'm on a 4.3* version of conda: 4.3.23

Corey-Zumar replied:
Ah, that seems to be the latest version. Will roll back and test with something earlier

Corey-Zumar reviewed (changes requested):
Almost there! A few comments concerning method outputs. I'm of the opinion that all method calls issued from a `ClipperConnection` object should log some sort of output. 

Also, there seems to be an issue with deploying python functions (see comment).

dcrankshaw committed:
addressed review comments (mostly by adding more logging messages)

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/664/
Test FAILed.

dcrankshaw committed:
Minor tweaks

dcrankshaw committed:
don't need to mount tmp in unit tests docker container anymore

Corey-Zumar reviewed (approved):
LGTM. Confirmed that the python closure deployment issue is only present for recent versions of anaconda. Will merge once this is rebased on develop and tests pass.

dcrankshaw commented:
Sweet. Actually hold off on merging until I get a chance to update the
quickstart and tutorial. Will do that in a few minutes.

On Wed, Aug 30, 2017 at 2:56 PM, Corey Zumar <notifications@github.com>
wrote:

> *@Corey-Zumar* approved this pull request.
>
> LGTM. Confirmed that the python closure deployment issue is only present
> for recent versions of anaconda. Will merge once this is rebased on develop
> and tests pass.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucbrise/clipper/pull/251#pullrequestreview-59706657>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaV5J2P0MxeeheBy7_xtBVUduEiggDLks5sddqRgaJpZM4OWfWX>
> .
>

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/665/
Test FAILed.

dcrankshaw committed:
update quickstart and re-add mount tmp in ci script

dcrankshaw committed:
updated example client

dcrankshaw committed:
Updated tutorial

dcrankshaw committed:
documented model deployers

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/666/
Test FAILed.

dcrankshaw committed:
Remove build sklearn cifar docker image from build_docker_images

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/667/
Test FAILed.

dcrankshaw commented:
jenkins test this please

Corey-Zumar commented on clipper_admin/clipper_admin/deployers/python.py:
```diff
@@ -21,33 +21,53 @@ def create_endpoint(
+        and used purely for user annotations.
+    registry : str, optional
+        The Docker container registry to push the freshly built model to. Note
+        that if you are running Clipper on Kubernetes, this registry must be accesible
```
Nit: s\accesible\accessible\

dcrankshaw replied:
Done

Corey-Zumar commented on clipper_admin/docs/index.rst:
```diff
@@ -30,16 +30,106 @@ one for running Clipper locally directly on Docker, and one for running Clipper
+In addition, the function must only use pure Python code. More specifically, all of the state
+captured by the function will be pickled using `Cloudpickle <https://github.com/cloudpipe/cloudpickle>`_,
+so any state captured by the function must be able to be pickled. Most Python libraries that
+use C extensions create objects that cannot be pickled. This includes many common machine-learning
```
this should be `...use C extensions **to** create objects that cannot be pickled...`

dcrankshaw replied:
I think it's correct as is. It's not the most elegant sentence in the world, but i'm leaving it.

Corey-Zumar commented on clipper_admin/docs/index.rst:
```diff
@@ -30,16 +30,106 @@ one for running Clipper locally directly on Docker, and one for running Clipper
+captured by the function will be pickled using `Cloudpickle <https://github.com/cloudpipe/cloudpickle>`_,
+so any state captured by the function must be able to be pickled. Most Python libraries that
+use C extensions create objects that cannot be pickled. This includes many common machine-learning
+frameworks including PySpark, TensorFlow, PyTorch, and Caffe. You will have to create your own
```
s\including\such as\ (we just used "including" earlier in the sentence)

dcrankshaw replied:
done

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/668/
Test FAILed.

dcrankshaw committed:
Fixed pyspark tests

Corey-Zumar commented on examples/tutorial/tutorial_part_two.ipynb:
```diff
@@ -136,18 +158,21 @@
    "metadata": {},
    "source": [
-    "To use your newly deployed model to generate predictions, it needs to be linked to your Clipper application. Let Clipper know that your `\"cifar_demo\"` app should use the `\"birds_vs_planes_classifier\"` model to serve predictions.\n",
+    "To use your newly deployed model to generate predictions, it needs to be linked to your Clipper application. This tells Clipper to route requests for `\"cifar-demo\"` app to the `\"birds-vs-planes-classifier\"` model to make predictions.\n",
```
s\\...to route requests for clipper demo app...\\...to route requests for **the** clipper demo app...\

dcrankshaw replied:
Done

Corey-Zumar reviewed (approved):
LGTM. Confirmed that quick start, the example client, and the tutorial work. Just fix a couple typos and I'll merge once tests pass.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/669/
Test FAILed.

dcrankshaw committed:
fixed pyspark test

dcrankshaw committed:
addressed comments and fixed pyspark test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/670/
Test FAILed.

dcrankshaw committed:
fixed a k8s test problem

dcrankshaw committed:
mount ~/.docker in unit tests docker file

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/671/
Test FAILed.

dcrankshaw committed:
commented out k8s test for now

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/672/
Test PASSed.
